File: Pandacap.ActivityPub.InboxRequests/ActivityPubInboxRequestHandler.cs
using Newtonsoft.Json.Linq;
using Pandacap.ActivityPub.Inbox.Interfaces;
using Pandacap.ActivityPub.InboxRequests.Interfaces;
using Pandacap.ActivityPub.RemoteObjects.Interfaces;

namespace Pandacap.ActivityPub.InboxRequests
{
    public class ActivityPubInboxRequestHandler(
        IActivityPubRemoteActorService activityPubRemoteActorService,
        IActivityPubRemotePostService activityPubRemotePostService,
        IActivityPubInboxActionHandler activityPubInboxActionHandler) : IActivityPubInboxRequestHandler
    {
        private static readonly IEnumerable<JToken> Empty = [];

        public async Task ProcessVerifiedInboxMessageAsync(
            JToken expansionObj,
            string myActorId,
            CancellationToken cancellationToken)
        {
            // Find out which ActivityPub actor they say they are
            string actorId = expansionObj["https://www.w3.org/ns/activitystreams#actor"]![0]!["@id"]!.Value<string>()!;

            // Grab that actor's information
            var actor = await activityPubRemoteActorService.FetchActorAsync(actorId, cancellationToken);

            string type = expansionObj["@type"]![0]!.Value<string>()!;

            if (type == "https://www.w3.org/ns/activitystreams#Follow")
            {
                string activityId = expansionObj["@id"]!.Value<string>()!;

                string fActor = expansionObj["https://www.w3.org/ns/activitystreams#actor"]![0]!["@id"]!.Value<string>()!;
                string fObject = expansionObj["https://www.w3.org/ns/activitystreams#object"]![0]!["@id"]!.Value<string>()!;

                if (fActor == actor.Id && fObject == myActorId)
                    await activityPubInboxActionHandler.RecordFollowAsync(activityId, actor, cancellationToken);
            }
            else if (type == "https://www.w3.org/ns/activitystreams#Undo")
            {
                foreach (var objectToUndo in expansionObj["https://www.w3.org/ns/activitystreams#object"] ?? Empty)
                {
                    if ((objectToUndo["@type"] ?? Empty).Any(token => token.Value<string>() == "https://www.w3.org/ns/activitystreams#Follow"))
                    {
                        string fActor = objectToUndo["https://www.w3.org/ns/activitystreams#actor"]![0]!["@id"]!.Value<string>()!;
                        string fObject = objectToUndo["https://www.w3.org/ns/activitystreams#object"]![0]!["@id"]!.Value<string>()!;
                        if (fActor == actor.Id && fObject == myActorId)
                        {
                            await activityPubInboxActionHandler.EraseFollowAsync(fActor, cancellationToken);
                        }
                    }

                    string id = objectToUndo["@id"]!.Value<string>()!;

                    await activityPubInboxActionHandler.EraseInteractionAsync(id, cancellationToken);
                    await activityPubInboxActionHandler.EraseAnnouncementAsync(id, cancellationToken);
                }
            }
            else if (type == "https://www.w3.org/ns/activitystreams#Accept")
            {
                foreach (var obj in expansionObj["https://www.w3.org/ns/activitystreams#object"] ?? Empty)
                {
                    string followId = obj["@id"]!.Value<string>()!;
                    await activityPubInboxActionHandler.MarkFollowerAsync(actor.Id, followId, accepted: true, cancellationToken);
                }
            }
            else if (type == "https://www.w3.org/ns/activitystreams#Reject")
            {
                foreach (var obj in expansionObj["https://www.w3.org/ns/activitystreams#object"] ?? Empty)
                {
                    string followId = obj["@id"]!.Value<string>()!;
                    await activityPubInboxActionHandler.MarkFollowerAsync(actor.Id, followId, accepted: false, cancellationToken);
                }
            }
            else if (type == "https://www.w3.org/ns/activitystreams#Like"
                || type == "https://www.w3.org/ns/activitystreams#Dislike"
                || type == "https://www.w3.org/ns/activitystreams#Flag"
                || type == "https://www.w3.org/ns/activitystreams#Listen"
                || type == "https://www.w3.org/ns/activitystreams#Read"
                || type == "https://www.w3.org/ns/activitystreams#View"
                || type == "https://www.w3.org/ns/activitystreams#Announce"
                || type == "https://ns.mia.jetzt/as#Bite")
            {
                foreach (string interactedWithId in activityPubRemotePostService.GetAnnouncementSubjectIds(expansionObj))
                {
                    if (type == "https://www.w3.org/ns/activitystreams#Announce")
                    {
                        await activityPubInboxActionHandler.RecordAnnouncementAsync(
                            actor,
                            expansionObj["@id"]!.Value<string>()!,
                            interactedWithId,
                            cancellationToken);
                    }

                    if (Uri.TryCreate(interactedWithId, UriKind.Absolute, out Uri? uri)
                        && uri != null
                        && Uri.TryCreate(myActorId, UriKind.Absolute, out Uri? me)
                        && me != null
                        && uri.Host == me.Host)
                    {
                        await activityPubInboxActionHandler.RecordInteractionAsync(
                            expansionObj["@id"]!.Value<string>()!,
                            interactedWithId,
                            actor.Id,
                            type,
                            cancellationToken);
                    }
                }
            }
            else if (type == "https://www.w3.org/ns/activitystreams#Create")
            {
                foreach (var obj in expansionObj["https://www.w3.org/ns/activitystreams#object"] ?? Empty)
                {
                    var remotePost = await activityPubRemotePostService.ParseExpandedObjectAsync(obj, cancellationToken);
                    await activityPubInboxActionHandler.RecordPostAsync(actor, remotePost, cancellationToken);
                }
            }
            else if (type == "https://www.w3.org/ns/activitystreams#Update")
            {
                foreach (var obj in expansionObj["https://www.w3.org/ns/activitystreams#object"] ?? Empty)
                {
                    string postId = obj["@id"]!.Value<string>()!;
                    string postType = obj["@type"]!.Value<string>()!;

                    if (postType == "Person" && postId == actor.Id)
                    {
                        await activityPubInboxActionHandler.UpdateRemoteActorAsync(actor, cancellationToken);
                    }

                    if (await activityPubInboxActionHandler.IsPostKnownAsync(postId, cancellationToken))
                    {
                        var remotePost = await activityPubRemotePostService.ParseExpandedObjectAsync(obj, cancellationToken);
                        await activityPubInboxActionHandler.UpdatePostAsync(actor, remotePost, cancellationToken);
                    }
                }
            }
            else if (type == "https://www.w3.org/ns/activitystreams#Delete")
            {
                foreach (var deletedObject in expansionObj["https://www.w3.org/ns/activitystreams#object"] ?? Empty)
                {
                    string deletedObjectId = deletedObject["@id"]!.Value<string>()!;

                    await activityPubInboxActionHandler.ErasePostAsync(actor.Id, deletedObjectId, cancellationToken);
                }
            }
        }
    }
}

File: Pandacap.ActivityPub.Services/ActivityPubInboxRequestHandler.fs
namespace Pandacap.ActivityPub.Services

open System
open Newtonsoft.Json.Linq
open Pandacap.ActivityPub.Inbox.Interfaces
open Pandacap.ActivityPub.RemoteObjects.Interfaces
open Pandacap.ActivityPub.Services.Interfaces

type internal ActivityPubInboxRequestHandler(
    activityPubRemoteActorService: IActivityPubRemoteActorService,
    activityPubRemotePostService: IActivityPubRemotePostService,
    activityPubInboxActionHandler: IActivityPubInboxActionHandler
) =
    interface IActivityPubInboxRequestHandler with
        member _.ProcessVerifiedInboxMessageAsync(expansionObj, myActorId, cancellationToken) = task {
            let expandedActorObject = Seq.exactlyOne expansionObj["https://www.w3.org/ns/activitystreams#actor"]
            let actorId = (expandedActorObject["@id"]).Value<string>()

            let! actor = activityPubRemoteActorService.FetchActorAsync(actorId, cancellationToken)
            if actor.Id = actorId then
                let activityId = (expansionObj["@id"]).Value<string>()
                let activityType = (Seq.exactlyOne expansionObj["@type"]).Value<string>()

                let objects = expansionObj["https://www.w3.org/ns/activitystreams#object"]

                let getId (object: JToken) = object["@id"].Value<string>()
                let getTypes (object: JToken) = seq {
                    if not (isNull object["@type"]) then
                        for token in object["@type"] do
                            token.Value<string>()
                }

                match activityType with
                | "https://www.w3.org/ns/activitystreams#Follow" ->
                    let target = Seq.exactlyOne objects
                    let targetId = getId target

                    if targetId = myActorId then
                        do! activityPubInboxActionHandler.RecordFollowAsync(activityId, actor, cancellationToken)

                | "https://www.w3.org/ns/activitystreams#Undo" ->
                    for object in objects do
                        do! activityPubInboxActionHandler.EraseAnnouncementAsync(getId object, actorId, cancellationToken)
                        do! activityPubInboxActionHandler.EraseInteractionAsync(getId object, actorId, cancellationToken)

                        let types = set (seq {
                            if not (isNull object["@type"]) then
                                for token in object["@type"] do
                                    token.Value<string>()
                        })

                        if types.Contains("https://www.w3.org/ns/activitystreams#Follow") then
                            let source = Seq.exactlyOne object["https://www.w3.org/ns/activitystreams#actor"]
                            let target = Seq.exactlyOne object["https://www.w3.org/ns/activitystreams#object"]

                            let sourceId = getId source
                            let targetId = getId target

                            if sourceId = actorId && targetId = myActorId then
                                do! activityPubInboxActionHandler.EraseFollowAsync(actorId, cancellationToken)

                | "https://www.w3.org/ns/activitystreams#Accept" ->
                    for object in objects do
                        do! activityPubInboxActionHandler.MarkFollowerAsync(getId object, actorId, true, cancellationToken)

                | "https://www.w3.org/ns/activitystreams#Reject" ->
                    for object in objects do
                        do! activityPubInboxActionHandler.MarkFollowerAsync(getId object, actorId, false, cancellationToken)

                | "https://www.w3.org/ns/activitystreams#Like"
                | "https://www.w3.org/ns/activitystreams#Dislike"
                | "https://www.w3.org/ns/activitystreams#Flag"
                | "https://www.w3.org/ns/activitystreams#Listen"
                | "https://www.w3.org/ns/activitystreams#Read"
                | "https://www.w3.org/ns/activitystreams#View"
                | "https://www.w3.org/ns/activitystreams#Announce"
                | "https://ns.mia.jetzt/as#Bite" ->
                    let tryParseUri str =
                        match Uri.TryCreate(str, UriKind.Absolute) with
                        | true, uri -> Some uri
                        | false, _ -> None

                    for targetObjectId in activityPubRemotePostService.GetAnnouncementSubjectIds(expansionObj) do
                        match tryParseUri targetObjectId, tryParseUri myActorId with
                        | (Some uri, Some me) when uri.Host = me.Host ->
                            do! activityPubInboxActionHandler.RecordInteractionAsync(activityId, targetObjectId, actorId, activityType, cancellationToken)
                        | _ -> ()

                        if activityType = "https://www.w3.org/ns/activitystreams#Announce" then
                            do! activityPubInboxActionHandler.RecordAnnouncementAsync(actor, activityId, targetObjectId, cancellationToken)

                | "https://www.w3.org/ns/activitystreams#Create" ->
                    for object in objects do
                        let! remotePost = activityPubRemotePostService.ParseExpandedObjectAsync(object, cancellationToken)
                        do! activityPubInboxActionHandler.RecordPostAsync(actor, remotePost, cancellationToken)

                | "https://www.w3.org/ns/activitystreams#Update" ->
                    for object in objects do
                        if getTypes object |> Seq.contains("https://www.w3.org/ns/activitystreams#Person") && getId object = actorId then
                            do! activityPubInboxActionHandler.UpdateRemoteActorAsync(actor, cancellationToken)

                        let! known = activityPubInboxActionHandler.IsPostKnownAsync(getId object, cancellationToken)
                        if known then
                            let! remotePost = activityPubRemotePostService.ParseExpandedObjectAsync(object, cancellationToken)
                            do! activityPubInboxActionHandler.UpdatePostAsync(actor, remotePost, cancellationToken)

                | "https://www.w3.org/ns/activitystreams#Delete" ->
                    for object in objects do
                        let! known = activityPubInboxActionHandler.IsPostKnownAsync(getId object, cancellationToken)
                        if known then
                            do! activityPubInboxActionHandler.ErasePostAsync(actorId, getId object, cancellationToken)

                | _ -> ()

            ()
        }
